dftimewolf/lib/processors/gcp_cloud_resource_tree.py

```diff
@@ -0,0 +1,1002 @@
+      query_filter = query_filter + \
+          f' AND (resource.labels.instance_id="{resource_id}" OR resource.labels.image_id="{resource_id}")'
+
+    log_messages: List[Dict[str,
+                            Any]] = gcl.ExecuteQuery(qfilter=[query_filter])
+
+    return log_messages
+
+  def _ParseLogMessagesFromFileContainer(
+      self, file_container: containers.File) -> None:
+    """Parses GCP Cloud log messages supplied in the file container and fill
+    the self._resource_dict with the result.
+
+    Args:
+      file_container: file container
+    """
+    if not file_container:
+      self.logger.error('File container is null')
+      return
+
+    if not file_container.path:
+      self.logger.error('File container path is null or empty')
+
+    with open(file_container.path, 'r') as input_file:
+      log_messages = []
+      file_content = input_file.readline()
+
+      if not file_content:
+        self.logger.error(f'The supplied file {file_container.path} is empty')
+        return
+
+      while file_content:
+        log_messages.append(json.loads(file_content))
+        file_content = input_file.readline()
+
+      self._ParseLogMessages(log_messages)
+
+  def _ParseLogMessages(self, log_messages: List[Dict[str, Any]]) -> None:
+    """Parses supplied GCP Cloud log messages and fills self._resource_dict with
+    the result.
+
+    Args:
+      log_messages: list of log messages
+    """
+    for log_message in log_messages:
+
+      proto_payload: Dict[str, Any] = log_message.get('protoPayload', {})
+
+      if not proto_payload:
+        continue
+
+      request_metadata = proto_payload.get('requestMetadata')
+      request = proto_payload.get('request')
+      response = proto_payload.get('response')
+      if not request or not response:
+        continue
+
+      log_message_type = request.get('@type').split('/')[
+          -1]  # Example: @type: "type.googleapis.com/compute.instances.insert"
+
+      # Parse logs for supported resource types
+      if log_message_type.startswith(
+          ('compute.instances', 'compute.disks', 'compute.machineImages',
+           'compute.image', 'compute.instanceTemplates', 'compute.snapshots')):
+        # Only parse 'operation' responses, skipping errors
+        if response.get('@type').split('/')[-1] != 'operation':
+          continue
+
+        # Check if a resource with the same ID already exist in the
+        # self._resources_dict dictionary
+        resource = self.resources_dict.get(response.get('targetId'))
+
+        if not resource:
+          resource = Resource()
+          resource.id = response.get('targetId')
+
+        # compute.disks.createSnapshot is a special case where the
+        # "resourceName" is just the name and not the full name with the
+        # project, zone, type and name
+        if log_message_type.startswith('compute.disks.createSnapshot'):
+          resource = Resource()
+          source_disk_id = response.get('targetId')
```

Comment:
move `source_disk_id`, `resource.project_id` and `resource.zone` value assignment to the else branch of the next if condition, these will be overridden with the matched_resource in case it is not empty, thus no need to assign them before the if

Reply:
Done